### PR TITLE
Add a HAProxy filebeat module.

### PR DESCRIPTION
**Pull request by Tethik**
I created a new module for parsing HAProxy http logs via filebeat. Originally developed on version 6.3 of elasticsearch/filebeat.

Feedback would be appreciated, please let me know if there's anything missing.

### CODE REVIEW

File: filebeat/module/haproxy/_meta/config.yml [hunk 1]
```diff
@@ -0,0 +1,8 @@
+- module: haproxy
+  # All logs
+  {fileset}:
```
Comment (jsoriano) [inline filebeat/module/haproxy/_meta/config.yml]: This should be replaced by the name of the fileset, I guess `http` in this case.

File: filebeat/docs/fields.asciidoc [hunk 2]
```diff
@@ -1435,6 +1436,339 @@ example: doc
...
+*`haproxy.http.time_duration`*::
++
+--
+type: long
+
+time_duration is the time the request remained active in haproxy, which is the total time in milliseconds elapsed between the first byte of the request was received and the last byte of response was sent.
+
+--
+
+*`haproxy.http.status_code`*::
```
Comment (jsoriano) [inline filebeat/docs/fields.asciidoc]: Could you follow `ECS` in fields naming if possible? for example [http response fields](https://github.com/elastic/ecs#http) could be at the root level under `http.response...`.

File: filebeat/module/haproxy/http/_meta/fields.yml [hunk 3]
```diff
@@ -0,0 +1,148 @@
...
+  - name: feconn
+    description: feconn is the total number of concurrent connections on the frontend when the session was logged.
+    
+    type: long
+  - name: beconn
+    description: beconn is the total number of concurrent connections handled by the backend when the session was logged.
+    
+    type: long
+  - name: srvconn
+    description: srv_conn is the total number of concurrent connections still active on the server when the session was logged.
```
Comment (jsoriano) [inline filebeat/module/haproxy/http/_meta/fields.yml]: These field names are a bit cryptic, maybe we could have a `connections` object with different counts, so these fields are named `connections.active`, `connections.frontend`, `connections.backend`, `connections.server`, wdyt?

File: filebeat/module/haproxy/_meta/config.yml [hunk 4]
```diff
@@ -0,0 +1,8 @@
...
+    enabled: true
+
+    # Set custom paths for the log files. If left empty,
+    # Filebeat will choose the paths depending on your OS.
+    #var.paths:
```
Comment (jsoriano) [inline filebeat/module/haproxy/_meta/config.yml]: Add here also a reference to `var.input`

File: filebeat/docs/modules/haproxy.asciidoc [hunk 5]
```diff
@@ -0,0 +1,62 @@
...
+This module comes with a sample dashboard showing geolocation, distribution of requests between backends and frontends,
+and status codes over time. For example:
+
+[role="screenshot"]
+image::./images/kibana-haproxy-overview.png[]
+
+include::../include/configuring-intro.asciidoc[]
+
+The module is by default configured to run via syslog on port 9001. However
+it can also be configured to read from a file path. See the following example.
```
Comment (jsoriano) [inline filebeat/docs/modules/haproxy.asciidoc]: Nice addition to be able to use syslog and file.

File: filebeat/docs/modules/haproxy.asciidoc [hunk 6]
```diff
@@ -0,0 +1,62 @@
...
+
+This module is not available for Windows.
+
+include::../include/running-modules.asciidoc[]
+
+[float]
+=== Example dashboard
+
+This module comes with a sample dashboard showing geolocation, distribution of requests between backends and frontends,
+and status codes over time. For example:
```
Comment (jsoriano) [inline filebeat/docs/modules/haproxy.asciidoc]: Thanks for the dashboard! :smiley:

File: filebeat/docs/modules/haproxy.asciidoc [hunk 7]
```diff
@@ -0,0 +1,62 @@
...
+    var.input: "file"
+-----
+
+:fileset_ex: http
+
+include::../include/config-option-intro.asciidoc[]
+
+
+[float]
+==== `{fileset}` log fileset settings
```
Comment (jsoriano) [inline filebeat/docs/modules/haproxy.asciidoc]: Replace `{fileset}` also here

File: filebeat/module/haproxy/_meta/config.yml (hunk 4 above)
Comment (Tethik) [reply filebeat/module/haproxy/_meta/config.yml]: :+1:

File: filebeat/module/haproxy/_meta/config.yml (hunk 1 above)
Comment (Tethik) [reply filebeat/module/haproxy/_meta/config.yml]: :+1:

File: filebeat/docs/modules/haproxy.asciidoc (hunk 7 above)
Comment (Tethik) [reply filebeat/docs/modules/haproxy.asciidoc]: :+1:

File: filebeat/docs/fields.asciidoc (hunk 2 above)
Comment (Tethik) [reply filebeat/docs/fields.asciidoc]: Will do.

File: filebeat/module/haproxy/http/_meta/fields.yml (hunk 3 above)
Comment (Tethik) [reply filebeat/module/haproxy/http/_meta/fields.yml]: I copied the names of the fields directly from the HAProxy documentation
https://cbonte.github.io/haproxy-dconv/1.7/configuration.html#8.2.3

I don't mind changing them though.

File: filebeat/docs/modules/haproxy.asciidoc (hunk 5 above)
Comment (Tethik) [reply filebeat/docs/modules/haproxy.asciidoc]: Thanks. I started out with a normal file log but setting it up in HAProxy requires a bit of effort. Going via syslog was easier as HAProxy has that support built in. I intend to document the HAProxy configuration necessary for using the module too.

File: filebeat/module/haproxy/http/_meta/fields.yml (hunk 3 above)
Comment (Tethik) [reply filebeat/module/haproxy/http/_meta/fields.yml]: Would you move them up from the fileset though? e.g. `haproxy.connections.*` as opposed to `haproxy.http.connections.*`

Comment (jsoriano) [reply filebeat/module/haproxy/http/_meta/fields.yml]: I think we could leave them directly on `haproxy.connections`, so if at some moment we have also a TCP fileset they can be aggregated more easily.

Comment (jsoriano) [reply filebeat/module/haproxy/http/_meta/fields.yml]: Good point, I also think it would also make sense to move these fields to the module level.

**Comment by Tethik** [comment]
Pushed some changes now, but I still need to update the dashboard and test the changes. Hope I can do so later tonight.

File: filebeat/module/haproxy/http/test/haproxy.log-expected.json [hunk 8]
```diff
@@ -0,0 +1,61 @@
+{
```
Comment (jsoriano) [inline filebeat/module/haproxy/http/test/haproxy.log-expected.json]: This file should contain only the list of events (each event being the value of `_source`)

File: filebeat/module/haproxy/http/test/haproxy.log-expected.json [hunk 9]
```diff
@@ -0,0 +1,61 @@
...
+                "geoip": {
+                  "continent_name": "Europe",
+                  "city_name": "Wildau",
+                  "country_iso_code": "DE",
+                  "region_name": "Brandenburg",
+                  "location": {
+                    "lon": 13.6333,
+                    "lat": 52.3167
+                  }
+                },
```
Comment (jsoriano) [inline filebeat/module/haproxy/http/test/haproxy.log-expected.json]: Be careful with exposing personal data in events :slightly_smiling_face:

File: filebeat/modules.d/haproxy.yml.disabled [hunk 10]
```diff
@@ -0,0 +1,8 @@
+- module: haproxy
+  # All logs
+  {fileset}:
```
Comment (jsoriano) [inline filebeat/modules.d/haproxy.yml.disabled]: `http` also here

File: filebeat/module/haproxy/http/manifest.yml [hunk 11]
```diff
@@ -0,0 +1,13 @@
+module_version: 1.0
+
+var:
+  - name: paths
+    default:
+      - /var/log/haproxy.log
+  - name: syslog_port
+    default: 9001
+  - name: input
+    default: syslog
```
Comment (jsoriano) [inline filebeat/module/haproxy/http/manifest.yml]: Being syslog the default is going to be tricky for the `test_modules` test as it instantiates filebeat with the default configurations.
@kvch @ph is there a way to provide a different config than the default for the `test_modules`?

**Comment by Tethik** [comment]
Still working on the dashboard fyi, quite busy these coming days though so it might take a while.

File: filebeat/filebeat.reference.yml [hunk 12]
```diff
@@ -116,6 +116,16 @@ filebeat.modules:
     # Filebeat will choose the paths depending on your OS.
     #var.paths:
 
+#------------------------------- haproxy Module ------------------------------
+- module: haproxy
+  # All logs
+  {fileset}:
```
Comment (jsoriano) [inline filebeat/filebeat.reference.yml]: Another placeholder here, remember to run also `make update` after modifying these files.

**Comment by Tethik** [comment]
Dashboard remade and tests scrubbed.

Here's also a sample haproxy config that should work to send logs to this module. 
https://gist.github.com/Tethik/b458036e734f5ec78bb362bba53e9d1c

File: filebeat/module/haproxy/http/manifest.yml (hunk 11 above)
Comment (kvch) [reply filebeat/module/haproxy/http/manifest.yml]: `filebeat/tests/system/config/filebeat.yml.j2` is the template used during running system tests. This needs to be extended with syslog input.

Comment (kvch) [reply filebeat/module/haproxy/http/manifest.yml]: The current system test framework also needs to be extended to be able to send logs over TCP and UDP to Filebeat. Maybe running netcat with proper config does the trick of feeding Filebeat with input.

Comment (jsoriano) [reply filebeat/module/haproxy/http/manifest.yml]: Thanks, I'll take a look to this.

**Comment by jsoriano** [comment]
@Tethik thanks for your work on this, I'm going to take a look on how we can extend our tests for syslog-based modules, I'll keep you updated.

**Comment by Tethik** [comment]
Glad to be able to contribute. I hope to add some more modules in the near future too :gift:

**Comment by jsoriano** [comment]
@Tethik we have merged #8140, that will force the use of file input for tests, please update your branch with master to check your test files.

**Comment by Tethik** [comment]
Done, I rebased to the latest master. Not sure how I'm supposed to run the tests though.

**Comment by jsoriano** [comment]
To run the tests locally, you can run `INTEGRATION_TESTS=1 make system-tests` from the `filebeat` directory.
This takes a while as it runs all system tests, if you want to run only the tests with the haproxy log files, you will need first a running elasticsearch, you can run it with `make start-environment`, and then:
```
make python-env
source build/python-env/bin/activate
make filebeat.test
INTEGRATION_TESTS=1 TEST_MODULES=haproxy BEAT_STRICT_PERMS=false ES_HOST=<easticsearch ip> nosetest tests/system/test_module.py
```
You can check the logs for more info in `build/system-tests`.

File: filebeat/module/haproxy/http/test/haproxy.log [hunk 13]
```diff
@@ -0,0 +1 @@
+Jul 30 09:03:52 localhost haproxy[32450]: 1.2.3.4:38862 [30/Jul/2018:09:03:52.726] incoming~ docs_microservice/docs 0/0/1/0/2 304 168 - - ---- 6/6/0/0/0 0/0 {docs.example.internal||} {|||} "GET /component---src-pages-index-js-4b15624544f97cf0bb8f.js HTTP/1.1"
```
Comment (jsoriano) [inline filebeat/module/haproxy/http/test/haproxy.log]: This file may need a new line at the end to ensure that this is read on tests.

File: filebeat/module/haproxy/http/test/haproxy.log-expected.json [hunk 14]
```diff
@@ -0,0 +1,41 @@
...
+        "haproxy.client_ip": "1.2.3.4",
+        "haproxy.time_backend_connect": 1,
+        "haproxy.connections.server": 0,
+        "haproxy.connections.retries": 0,
+        "haproxy.connections.active": 6,
+        "haproxy.connections.backend": 0,
+        "haproxy.connections.frontend": 6,
+        "@timestamp": "2018-07-30T09:03:52.726Z",
+        "message": "Jul 30 09:03:52 localhost haproxy[32450]: 1.2.3.4:38862 [30/Jul/2018:09:03:52.726] incoming~ docs_microservice/docs 0/0/1/0/2 304 168 - - ---- 6/6/0/0/0 0/0 {docs.example.internal||} {|||} \"GET /component---src-pages-index-js-4b15624544f97cf0bb8f.js HTTP/1.1\""
+    }
```
Comment (jsoriano) [inline filebeat/module/haproxy/http/test/haproxy.log-expected.json]: These fields are missing:
```
        "input.type": "log",
        "prospector.type": "log",
        "fileset.module": "haproxy",
        "fileset.name": "http",
        "haproxy.geoip.region_iso_code": "US-WA",
        "offset": 0,
```

File: filebeat/module/haproxy/http/test/haproxy.log (hunk 13 above)
Comment (Tethik) [reply filebeat/module/haproxy/http/test/haproxy.log]: Pushed the newline. No change though

Comment (Tethik) [reply filebeat/module/haproxy/http/test/haproxy.log]: Ah sorry, been trying to run the tests locally but the environment is not cooperating with me. (I get different failures from the other unrelated tests too)

**Comment by jsoriano** [comment]
jenkins, test this please

**Comment by jsoriano** [comment]
jenkins, test this again please

**Comment by jsoriano** [comment]
To be backported to 6.x along with #8215

**Pull request merged**